src/python/pants/engine/target.py

```diff
@@ -22,25 +24,55 @@ def __repr__(self) -> str:
 class PrimitiveField(Field, metaclass=ABCMeta):
     """A Field that does not need the engine in order to be hydrated.
 
-    This should be subclassed by the majority of fields.
+    This should be used by the majority of fields.
+
+    Subclasses must implement `hydrate()` to convert `self.raw_value` into `self.value`. This
+    hydration and/or validation happens eagerly in the constructor. If the hydration is
+    particularly expensive, use `AsyncField` instead to get the benefits of the engine's caching.
+
+    Subclasses should also override the type hints for `raw_value` and `value` to be more precise
+    than `Any`.
+
+    Example:
+
+        class ZipSafe(PrimitiveField):
+            alias: ClassVar = "zip_safe"
+            raw_value: Optional[bool]
+            value: bool
+
+            def hydrate(self) -> bool:
+                if self.raw_value is None:
+                    return True
+                return self.raw_value
     """
 
-    def __str__(self) -> str:
-        return f"{self.alias}={self.value}"
+    value: Any
+
+    @final
+    def __init__(self, raw_value: Optional[Any]) -> None:
+        self.raw_value = raw_value
+        self.value = self.hydrate()
 
-    @memoized_property
     @abstractmethod
-    def value(self) -> Any:
-        """Get the field's value.
+    def hydrate(self) -> Any:
```

Comment:
Originally, I didn't have this method and thought "Every `Field` has its own custom `__init__()`!". 

But, this got messy really quickly, e.g. that the `Field` would have to be careful to properly set `self.raw_value`. It also became hard to subclass. `__init__()` mutates the instance and returns `None`, whereas this method returns the actual value as a pure function. By making this be a pure function, we can use this idiom:

```python
 def hydrate(self) -> List[str]:
             hydrated_val = super().hydrate()
             if len(hydrated_val) != 3:
                 raise ValueError("Bad!")
             return hydrated_val
```